### PR TITLE
feat: Add Windows ipconfig parser to jc

### LINKED ISSUE #594: Parser Request: "ipconfig /all" (Windows)
I'm finding myself in need of an ipconfig parser and think this is likely others might find useful. Looking at the output, it looks like it might be rather simple to parse reliably compared to some of the other open parser requests I was looking at. 

Here is a sample `ipconfig /all` for discussion.

```Windows IP Configuration

   Host Name . . . . . . . . . . . . : ABC-PC1
   Primary Dns Suffix  . . . . . . . : abc.corp
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : abc.corp

Unknown adapter Tailscale:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Tailscale Tunnel
   Physical Address. . . . . . . . . :
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::d8ff:5060:f230:9168%5(Preferred)
   Autoconfiguration IPv4 Address. . : 169.254.83.107(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.0.0
   Default Gateway . . . . . . . . . :
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Disabled

Ethernet adapter Ethernet0:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Intel(R) 82574L Gigabit Network Connection
   Physical Address. . . . . . . . . : 00-0C-2A-0A-7A-D4
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::47ac:e226:3b46:a740%7(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.22.108(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : Tuesday, June 11, 2024 12:19:26 PM
   Lease Expires . . . . . . . . . . : Friday, September 20, 2024 4:25:16 AM
   Default Gateway . . . . . . . . . : 192.168.22.1
   DHCP Server . . . . . . . . . . . : 192.168.22.1
   DHCPv6 IAID . . . . . . . . . . . : 100634309
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-29-4D-7E-49-00-0A-22-08-24-BD
   DNS Servers . . . . . . . . . . . : 192.168.22.151
   NetBIOS over Tcpip. . . . . . . . : Enabled
   ```

### PR DESCRIPTION
**Pull request by joehacksalot**
This PR introduces a new parser for the Windows ipconfig command to address issue #594.

Tested on:
- Windows XP
- Windows 7
- Windows 10
- Windows 11
- Windows Server 2008
- Windows Server 2016

Also, manually modified the configuration of a windows VM to test for a multiple entries for: 
- IPv6 addresses
- Link-local IPv6 addresses
- Temporary IPv6 addresses
- Default gateways (one IPv4 + one IPv6)
- DNS servers

Notes (copied from ipconfig.py:85):
```
[0] - 'lease_expires' and 'lease_obtained' are parsed to ISO8601 format date strings. if the value was unable 
        to be parsed by datetime, the fields will be in their raw form
[1] - 'autoconfigured' under 'ipv4_address' is only providing indication if the ipv4 address was labeled as
        "Autoconfiguration IPv4 Address" vs "IPv4 Address". It does not infer any information from other fields
[2] - Windows XP uses 'IP Address' instead of 'IPv4 Address'. Both values are parsed to the 'ipv4_address' 
        object for consistency
 ```

Closes #594

### CODE REVIEW

**Comment by kellyjonbrazil** [comment]
Nice contribution, thanks! I'll take a closer look in the next few days.

File: jc/parsers/ipconfig.py [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+    if key == "host_name":
+        result["host_name"] = value
+    elif key == "primary_dns_suffix":
+        result["primary_dns_suffix"] = value
+    elif key == "node_type":
+        result["node_type"] = value
+    elif key == "ip_routing_enabled":
+        result["ip_routing_enabled"] = value
+    elif key == "wins_proxy_enabled":
+        result["wins_proxy_enabled"] = value
```
Comment (kellyjonbrazil) [inline jc/parsers/ipconfig.py:675]: Might be able to simplify with something like:
```
if key in ['host_name', 'primary_dns_suffix', ...]:
    result[key] = value
```

File: jc/parsers/ipconfig.py [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+    elif key == "netbios_over_tcpip":
+        adapter["netbios_over_tcpip"] = value
+    elif key == "dhcp_server":
+        adapter["dhcp_server"] = value
+    elif key == "lease_obtained":
+        adapter["lease_obtained"] = value
+    elif key == "lease_expires":
+        adapter["lease_expires"] = value
+    elif key == "primary_wins_server":
+        adapter["primary_wins_server"] = value
```
Comment (kellyjonbrazil) [inline jc/parsers/ipconfig.py:711]: Might be able to simplify with something like:
```
if key in ['connection_specific_dns_suffix', 'media_state', ...]:
    adapter[key] = value
```

**Review by kellyjonbrazil — COMMENTED** [review]
Nice parser! On my first pass just saw a couple sections that could be shortened to clarify intent. I see the tests are failing on an IPv6 address with "Preferred" being lower-case or something.

**Comment by joehacksalot** [comment]
That was a last minute change to restore status text to source casing (previously had it all lowercase). I'll update the tests shortly.

**Comment by joehacksalot** [comment]
Ok. Updates are in.

File: jc/parsers/ipconfig.py [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+            adapter["autoconfiguration_enabled"] = (adapter["autoconfiguration_enabled"].lower() == "yes")
+        if "netbios_over_tcpip" in adapter and adapter["netbios_over_tcpip"] is not None:
+            adapter["netbios_over_tcpip"] = (adapter["netbios_over_tcpip"].lower() == "enabled")
+        if "lease_expires" in adapter and adapter["lease_expires"] is not None and adapter["lease_expires"] != "":
+            try:
+                adapter["lease_expires"] = datetime.strptime(adapter["lease_expires"], "%A, %B %d, %Y %I:%M:%S %p").isoformat()
+            except:
+                pass # Leave date in raw format if not parseable
+        if "lease_obtained" in adapter and adapter["lease_obtained"] is not None and adapter["lease_obtained"] != "":
+            try:
```
Comment (kellyjonbrazil) [inline jc/parsers/ipconfig.py:531]: I think what I'll do here is add this time format to the `jc.utils.timestamp()` class and then we can keep the original field and add epoch and iso timestamps if it parses.

Comment (joehacksalot) [reply jc/parsers/ipconfig.py:531]: I wasn't really sure what to do here with this one. I'm not using these timestamps but I wanted to leave them in a useful and easily consumable format for others to use. ISO8601 seemed to be a good option but I have no issues moving to something consistent with other parsers for this data type.

**Pull request merged**